lib/src/rules/function_expression_with_incorrect_types.dart

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../analyzer.dart';
+
+const _desc =
+    r'Use the parameter types expected by the target in function expression.';
+
+const _details = r'''
+
+Use the parameter types expected by the target in function expression.
+
+**BAD:**
+```
+f(void Function(int a) p) {}
+
+f((int? a) {});
+f((num a) {});
+```
+
+**GOOD:**
+```
+f(void Function(int a) p) {}
+
+f((a) {});
+f((int a) {});
+```
+
+''';
+
+class FunctionExpressionWithIncorrectTypes extends LintRule
+    implements NodeLintRule {
+  FunctionExpressionWithIncorrectTypes()
+      : super(
+            name: 'function_expression_with_incorrect_types',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addFunctionExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitFunctionExpression(FunctionExpression node) {
+    DartType? expectedType;
+    var parent = node.unParenthesized.parent;
+    if (parent is ArgumentList) {
+      var parentParent = parent.parent;
+      if (parentParent is MethodInvocation) {
+        final element = parentParent.staticInvokeType?.element;
+        if (element is FunctionElement) {
+          expectedType = element
+              .parameters[parent.arguments.indexOf(node.unParenthesized)].type;
+        }
+      }
+    } else if (parent is AssignmentExpression) {
+      expectedType = parent.writeType;
+    }
+
+    // exit if the case is not covered
+    if (expectedType == null || expectedType is! FunctionType) return;
+
+    // check type of each parameters
+    var currentType = node.staticType! as FunctionType;
+    var currentParams = currentType.parameters;
+    var expectedParams = expectedType.parameters;
+    for (var i = 0; i < min(expectedParams.length, currentParams.length); i++) {
```

Comment:
I'm guessing that this won't work for named parameters if the parameters are defined in different orders or if the closure allows named parameters that aren't in the expected type. For example
```
void g(void Function({String? s, int? i}) f) {}
void h() {
  g(({int? i, String? s}) {});
}
```

Reply:
Fixed.

Reply:
Not fixed at all as names for positionnal parameters can change... Sorry